chain-vote/src/tally.rs

```diff
@@ -61,20 +89,26 @@ impl EncryptedTally {
     /// Initialise a new tally with N different options. The `EncryptedTally` is computed using
     /// the additive homomorphic property of the elgamal `Ciphertext`s, and is therefore initialised
     /// with zero ciphertexts.
-    pub fn new(options: usize) -> Self {
+    pub fn new(options: usize, election_pk: ElectionPublicKey, crs: Crs) -> Self {
         let r = vec![Ciphertext::zero(); options];
-        EncryptedTally { r }
+        EncryptedTally {
+            r,
+            fingerprint: (&election_pk, &crs).into(),
+        }
     }
 
-    /// Add a submitted `ballot`, with a specific `weight` to the tally, if
-    /// the `ballot` contains a valid proof. If the proof is invalid, it will
-    /// panic. todo: maybe we want to handle these errors?
+    /// Add a submitted `ballot`, with a specific `weight` to the tally.
+    /// Remember that a vote is only valid for a specific election (i.e. pair of
+    /// election public key and crs), and trying to add a ballot validated for a
+    /// different one will result in a panic.
     ///
     /// Note that the encrypted vote needs to have the exact same number of
     /// options as the initialised tally, otherwise an assert will trigger.
     #[allow(clippy::ptr_arg)]
-    pub fn add(&mut self, vote: &EncryptedVote, weight: u64) {
-        for (ri, ci) in self.r.iter_mut().zip(vote.iter()) {
+    pub fn add(&mut self, ballot: &Ballot, weight: u64) {
+        assert_eq!(ballot.vote.len(), self.r.len());
+        assert_eq!(ballot.fingerprint, self.fingerprint);
```

Comment:
Those are already validated when I obtain a `Ballot` the first time. If those checks fail here it means that the programmer made some mistake and tried to apply a `Ballot` to a different `EncryptedTally` than the one it was verified against, and is a serious mistake.

I would have agreed with ad-hoc errors if this was the first time the vote was verified, but this PR is here to avoid that